manta-trusted-setup/src/groth16/ppot/hashing.rs

```diff
@@ -0,0 +1,226 @@
+    pub type SubCeremony = PerpetualPowersOfTauCeremony<PpotSerializer, POWERS>;
+
+    /// Checks a transition between challenge files.  Note that the appropriate challenge
+    /// and reponse files must belong to `manta-parameters`.
+    #[ignore] // NOTE: Adds `ignore` such that CI does NOT run this test while still allowing developers to test.
+    #[test]
+    fn verify_one_transition_test() {
+        use ark_std::{fs::OpenOptions, time::Instant};
+        use memmap::MmapOptions;
+
+        // Read first accumulator
+        println!("Reading accumulator from challenge file");
+        let now = Instant::now();
+        let reader = OpenOptions::new()
+            .read(true)
+            .open("../manta-parameters/data/ppot/challenge_0071.lfs") // TODO: This path doesn't work
+            .expect("unable open `./challenge` in this directory");
+        let challenge_map = unsafe {
+            MmapOptions::new()
+                .map(&reader)
+                .expect("unable to create a memory map for input")
+        };
+        let prev_accumulator =
+            read_subaccumulator::<SubCeremony>(&challenge_map, Compressed::No).unwrap();
+        println!("Read uncompressed accumulator in {:?}", now.elapsed());
+
+        // Read second accumulator
+        println!("Reading accumulator from challenge file");
+        let now = Instant::now();
+        let reader = OpenOptions::new()
+            .read(true)
+            .open("../manta-parameters/data/ppot/challenge_0072.lfs")
+            .expect("unable open `./challenge` in this directory");
+        let challenge_map = unsafe {
+            MmapOptions::new()
+                .map(&reader)
+                .expect("unable to create a memory map for input")
+        };
+        let next_accumulator =
+            read_subaccumulator::<SubCeremony>(&challenge_map, Compressed::No).unwrap();
+        println!("Read uncompressed accumulator in {:?}", now.elapsed());
+
+        // Load `response` file
+        println!("Reading accumulator from response file");
+        let reader = OpenOptions::new()
+            .read(true)
+            .open("../manta-parameters/data/ppot/response_0071.lfs")
+            .expect("unable open `./response` in this directory");
+        let response = unsafe {
+            MmapOptions::new()
+                .map(&reader)
+                .expect("unable to create a memory map for input")
+        };
+
+        let challenge_hash: [u8; 64] = into_array_unchecked(
+            response
+                .get(0..64)
+                .expect("Cannot read hash from header of response file"),
+        );
+        let proof: Proof<PpotCeremony> = read_kzg_proof(&response).unwrap();
+
+        let _prev = Accumulator::<SubCeremony>::verify_transform(
+            prev_accumulator,
+            next_accumulator,
+            challenge_hash,
+            proof.cast_to_subceremony(),
+        )
+        .unwrap();
+    }
+}
```

Comment:
Q: If you take the files from an external source, and considering that the hash takes a challenge which will, in principle, depend on the ceremony size, when we take the subceremony we have no control over the new hashes, right? So even if the transition from BIG_FILE_1 to BIG_FILE_2 satisfies the test, when we reduce it to SMALL_FILE_1 and SMALL_FILE_2, shouldn't it fail?

Reply:
Yep, that's right. The proof is tied to the hash of the file, so if you change the file you change the proof and nothing works. So what we do is we always hash the _full file_ even if we're only verifying a subset of the accumulator it contains. This is why we had to download all 10 Tb of data to verify PPoT even though we only use 1/500th of the actual accumulators.

In this test we use a shortcut: if you look [here](https://github.com/Manta-Network/manta-rs/blob/3dbf4b5faec4d4a0c7f7c83d4dcd01528fbd8589/manta-trusted-setup/src/groth16/ppot/hashing.rs#L211) you'll see that instead of hashing a file we actually read the hash from another file.  That's because `response_k` is _supposed_ to contain the hash of `challenge_k` at the beginning of the file.  Now it might not, but if you optimistically assume that it does then you can separate verification into checking hashes and checking transitions and do those checks separately.

Reply:
I see, all clear. Do you think we should hash our small files anyway and store the values somewhere? This way we'd have some protection against tampering that doesn't involve going to the external files and checking their hashes.